File: Python/219.Contains Duplicate II.py
#求是否存在nums[i]=nums[j]，且i与j的差不超过k
class Solution(object):
    #用字典存一下每个数最新出现的位置
    def containsNearbyDuplicate(self, nums, k):
        dic = {}
        for i in range(len(nums)):
            if dic.get(nums[i],-1)!=-1 and i-dic[nums[i]]<=k:
                return True
            dic[nums[i]] = i
        return False

    #蛮力法，超时
    def containsNearbyDuplicate1(self, nums, k):
        l = len(nums)
        if l<=1:    return False
        for i in range(l-1):
            r = l if i+k>=l else i+k
            for j in range(i+1,r):
                if nums[i]==nums[j]:
                    return True
        return False

s = Solution()
print(s.containsNearbyDuplicate([1,0,1,1],1))

File: C++/219.Contains Duplicate II.cpp
/*
判断数字中是否存在两个下标i和j，有nums[i]==nums[j]
而且|i-j|<=k
*/


#include <iostream>
#include <vector>
#include <map>
#include <math.h>
using namespace std;

class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        map<int,int> mymap;
        map<int,int>::iterator it;
        for(int i=0; i < nums.size(); ++i)
        {
        	it = mymap.find(nums[i]);
        	
        	if(it==mymap.end())
        	{	
        		mymap.emplace(nums[i],i);
        	}
        	else
        	{
        		if(fabs(i-it->second)<=k)	return true;
        		it->second = i;
        	}
        }
        return false;
    }
};

int main()
{
	Solution s;
	vector<int> nums;
	nums.push_back(1);
	nums.push_back(0);
	nums.push_back(1);
	nums.push_back(1);
	//nums.push_back(1);
	//nums.push_back(2);
	//nums.push_back(2);
	cout<<s.containsNearbyDuplicate(nums,1)<<endl;
}